### PR TITLE
feat: vapo notebooks and lib

### PR DESCRIPTION
**Pull request by inardini**
# Description
 This PR is about updating vapo notebooks and library. 

Thank you for opening a Pull Request!
Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [x] Follow the [`CONTRIBUTING` Guide](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/CONTRIBUTING.md).
- [x] You are listed as the author in your notebook or README file.
  - [x] Your account is listed in [`CODEOWNERS`](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/.github/CODEOWNERS) for the file(s).
- [x] Make your Pull Request title in the <https://www.conventionalcommits.org/> specification.
- [x] Ensure the tests and linter pass (Run `nox -s format` from the repository root to format).
- [x] Appropriate docs were updated (if necessary)

### CODE REVIEW

File: gemini/prompts/prompt_optimizer/vapo_lib.py [hunk 1]
```diff
@@ -660,3 +670,189 @@ def get_container(self) -> widgets.Output:
                 self.results_output,
             ]
         )
+
+
+def get_id(length: Union[int, None] = 8) -> str:
+    """Generate a uuid of a specified length (default=8)."""
+    if length is None:
+        length = 8
```
Comment (holtskinner) [inline gemini/prompts/prompt_optimizer/vapo_lib.py:678]:
```suggestion
def get_id(length: int = 8) -> str:
    """Generate a uuid of a specified length (default=8)."""
```
I don't think you need to have `None` as an optional argument here, this accomplishes the same thing.

File: gemini/prompts/prompt_optimizer/vapo_lib.py [hunk 2]
```diff
@@ -660,3 +670,189 @@ def get_container(self) -> widgets.Output:
...
+        result = subprocess.run(
+            ["gcloud", "auth", "print-identity-token", "-q"],
+            capture_output=True,
+            text=True,
+            check=True,
+        )
+        return result.stdout.strip()
+    except subprocess.CalledProcessError as e:
+        print(f"Error getting auth token: {e}")
+        return None
```
Comment (holtskinner) [inline gemini/prompts/prompt_optimizer/vapo_lib.py:694]: There should be a way to do this without running the gcloud cli command

File: gemini/prompts/prompt_optimizer/vapo_lib.py [hunk 3]
```diff
@@ -660,3 +670,189 @@ def get_container(self) -> widgets.Output:
...
+) -> GenerativeModel:
+    """Initialize a new model with configurable generation and safety settings."""
+
+    # Use default configurations if none are provided
+    if generation_config is None:
+        generation_config = {
+            "candidate_count": 1,
+            "max_output_tokens": 2048,
+            "temperature": 0.5,
+        }
```
Comment (holtskinner) [inline gemini/prompts/prompt_optimizer/vapo_lib.py:800]: Use the `GenerationConfig` class

File: gemini/prompts/prompt_optimizer/vapo_lib.py [hunk 4]
```diff
@@ -660,3 +670,189 @@ def get_container(self) -> widgets.Output:
...
+    # Iterate over the rows of the DataFrame
+    for _, row in df.iterrows():
+        for field in df.columns:
+            display(HTML(f"<span style='{header_style}'>{field.capitalize()}:</span>"))
+            display(HTML("<br>"))
+            value = row[field]
+            display(HTML(f"<span style='{base_style}'>{value}</span>"))
+            display(HTML("<br>"))
+
+        printed_samples += 1
```
Comment (holtskinner) [inline gemini/prompts/prompt_optimizer/vapo_lib.py:781]: Could you use the index value from `df.iterrows()` instead of a new variable?

**Review by holtskinner — CHANGES_REQUESTED** [review]

File: gemini/prompts/prompt_optimizer/vapo_lib.py (hunk 1 above)
Comment (inardini) [reply gemini/prompts/prompt_optimizer/vapo_lib.py:678]: done

File: gemini/prompts/prompt_optimizer/vapo_lib.py (hunk 2 above)
Comment (inardini) [reply gemini/prompts/prompt_optimizer/vapo_lib.py:694]: I agree. I tried the auth module of the google lib but I get some unexpected behaviour when I tried to call the cloud function I am using the token for.

File: gemini/prompts/prompt_optimizer/vapo_lib.py (hunk 3 above)
Comment (inardini) [reply gemini/prompts/prompt_optimizer/vapo_lib.py:800]: done

**Pull request merged**